### PR TITLE
Do not configure ip6tables at all if IPv6 is disabled

### PR DESCRIPTION
**Pull request by rlenglet**
Fixes https://github.com/istio/istio/issues/19380
Related to https://github.com/istio/cni/pull/172

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by rlenglet** [comment]
@clyang82 fyi

**Comment by clyang82** [comment]
@rlenglet Thanks for your quick response.

File: tools/istio-iptables/pkg/cmd/run.go [hunk 1]
```diff
@@ -436,7 +428,9 @@ func (iptConfigurator *IptablesConfigurator) run() {
 	}
 
 	iptConfigurator.handleInboundIpv4Rules(ipv4RangesInclude)
-	iptConfigurator.handleInboundIpv6Rules(ipv6RangesExclude, ipv6RangesInclude)
+	if iptConfigurator.cfg.EnableInboundIPv6s != nil {
```
Comment (abhide) [inline tools/istio-iptables/pkg/cmd/run.go]: I think we should this same check on L316 where `ip6tables-save` is invoked.

Comment (abhide) [reply tools/istio-iptables/pkg/cmd/run.go]: https://github.com/istio/istio/blob/master/tools/istio-iptables/pkg/cmd/run.go#L316

Comment (rlenglet) [reply tools/istio-iptables/pkg/cmd/run.go]: Done.

**Comment by rlenglet** [comment]
I had to refactor the code to inject the IPv6 support configuration instead of doing the detection direction in the rule generation code.
While doing that, I found and fixed the handling of the `ENVOY_USER` and `DISABLE_REDIRECTION_ON_LOCAL_LOOPBACK` environment variables. Those were simply always ignored. The effect is that if the `-u` flag was not passed, the UID and GID were set to `0,0` instead of `1337,0`. That broke VM support.

@abhide As a side-effect of this refactoring, the `Dependencies` interface is now cleaner. It only defines `Run*()` method. It can and should now be renamed to something more specific like `Runner`.

**Comment by rlenglet** [comment]
The `unit-tests_istio` test failure is legit.
The test script simulates IPv6 being enabled by setting the `STUB_IP` environment variable:
https://github.com/istio/istio/blob/6844697b61dc54caa1e20d01ca56264a4f769cc1/tests/scripts/istio-iptables-test.sh#L110-L112

The only place where this is used is in the `hostname` command mock:
https://github.com/istio/istio/blob/6844697b61dc54caa1e20d01ca56264a4f769cc1/tests/scripts/stubs/hostname#L18

The `istio-iptables.sh` _did_ use the `hostname` command to determine whether IPv6 is enabled. But the golang implementation has never used `hostname`. It has always relied on the IPv6 addresses assigned to the host:
https://github.com/istio/istio/blob/6844697b61dc54caa1e20d01ca56264a4f769cc1/tests/scripts/stubs/hostname#L18
https://github.com/istio/istio/blob/6844697b61dc54caa1e20d01ca56264a4f769cc1/tools/istio-iptables/pkg/cmd/run.go#L340-L348

So this test should have always failed.

I don't know how this test would have ever passed for the Golang implementation.
Maybe this test was never run somehow?

**Comment by rlenglet** [comment]
I found why the test wasn't failing.
The test is executing `istio-iptables` in dry-run mode:
https://github.com/istio/istio/blob/6844697b61dc54caa1e20d01ca56264a4f769cc1/tests/scripts/istio-iptables-test.sh#L61

In dry-run mode, the command was using a different implementation to determine the host IP address than the normal mode, which was using the `STUB_IP` environment variable:
https://github.com/istio/istio/blob/6844697b61dc54caa1e20d01ca56264a4f769cc1/tools/istio-iptables/pkg/cmd/run.go#L43-L47
https://github.com/istio/istio/blob/6844697b61dc54caa1e20d01ca56264a4f769cc1/tools/istio-iptables/pkg/dependencies/stub.go#L31-L34

So the dry-run mode was detecting differently whether IPv6 is enabled, so it had a different behavior from the non-dry-run mode and was generating different commands. That was a bug. The dry-run mode was using a mock meant for testing instead of a proper implementation. I fixed that bug in this PR.

File: tools/istio-iptables/pkg/builder/iptables_builder_impl.go [hunk 2]
```diff
@@ -110,11 +108,11 @@ func (rb *IptablesBuilderImpl) buildRules(command string, rules []*Rule) [][]str
 }
 
 func (rb *IptablesBuilderImpl) BuildV4() [][]string {
-	return rb.buildRules(dependencies.IPTABLES, rb.rules.rulesv4)
+	return rb.buildRules(constants.IPTABLES, rb.rules.rulesv4)
```
Comment (abhide) [inline tools/istio-iptables/pkg/builder/iptables_builder_impl.go]: My bad! Thanks for fixing!

File: tools/istio-iptables/pkg/cmd/root.go [hunk 3]
```diff
@@ -15,30 +15,49 @@
...
+	"istio.io/pkg/env"
 
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
 
 	"istio.io/pkg/log"
 )
 
+var (
+	envoyUserVar = env.RegisterStringVar(constants.EnvoyUser, "istio-proxy", "Envoy proxy username")
```
Comment (abhide) [inline tools/istio-iptables/pkg/cmd/root.go]: any way we can move this to `constructConfig`. Would like to keep reading config at a single place and also use `viper` as a standard way to read env vars

File: tools/istio-iptables/pkg/builder/iptables_builder_impl.go (hunk 2 above)
Comment (rlenglet) [reply tools/istio-iptables/pkg/builder/iptables_builder_impl.go]: Hey, no problem!

File: tools/istio-iptables/pkg/cmd/root.go (hunk 3 above)
Comment (rlenglet) [reply tools/istio-iptables/pkg/cmd/root.go]: Unfortunately, no. This API is meant to be used this way. See all other usages.
This function must be called at package init time, but at the same time there's no way to retrieve the variable after calling `Register*Var()`. So the value returned by the call must be stored in a variable somewhere. A global variable is the only way to use this API properly.

At least, now those variables will be documented in the HTML docs.

**Comment by rlenglet** [comment]
@istio/wg-test-and-release-maintainers Please the test changes for this PR. Those are changes to the iptables test golden files due do:

- Adding new output for debugging (`PROXY_GID`).
- Replacing the debug output for IPv6 (it's now a boolean).
- Fixing the default `ENVOY_UID` when none is specified in the command-line flags.

File: Makefile.core.mk [hunk 4]
```diff
@@ -437,8 +437,8 @@ security-test:
 common-test: build
 	go test ${T} ./pkg/...
 	go test ${T} ./tests/common/...
+	go test ${T} ./tools/istio-iptables/...
 	# Execute bash shell unit tests scripts
-	./tests/scripts/scripts_test.sh
 	./tests/scripts/istio-iptables-test.sh
```
Comment (howardjohn) [inline Makefile.core.mk]: in future we should make this `go` code. Even if its just calling out to shell, for consistency with tooling

File: Makefile.core.mk [hunk 5]
```diff
@@ -437,8 +437,8 @@ security-test:
 common-test: build
```
Comment (howardjohn) [inline Makefile.core.mk]: this target isn't run in CI, we call common-racetest. Now that I am thinking about it, we should probably just remove the racetest targets and make these ones pass `-race`

Comment (rlenglet) [reply Makefile.core.mk]: Do you want me to do it in this PR?

Comment (howardjohn) [reply Makefile.core.mk]: I was thinking this change was not tested as a result of this. If it is tested still we can do it separately

Comment (rlenglet) [reply Makefile.core.mk]: Yes, I ran `make common-test` by hand and made sure everything is passing.

**Comment by rlenglet** [comment]
The failure `e2e-simpleTests-distroless_istio` seems to be an environment failure:
```
2020-01-04T01:06:53.231092Z	info	Running command helm install --dry-run --debug /home/prow/go/src/istio.io/istio/install/kubernetes/helm/istio --name istio --values /home/prow/go/src/istio.io/istio/install/kubernetes/helm/istio/test-values/values-e2e.yaml --namespace istio-system --set global.mtls.enabled=true --set galley.enabled=true --set global.useMCP=true --set global.configValidation=true --set global.outboundTrafficPolicy.mode=ALLOW_ANY --set-string global.hub=istio-testing --set-string global.tag=istio-testing-distroless --set global.enableHelmTest=true
2020-01-04T01:06:53.524230Z	info	Command output: 
[debug] Created tunnel using local port: '37341'

[debug] SERVER: "127.0.0.1:37341"

[debug] Original chart version: ""
[debug] CHART PATH: /home/prow/go/src/istio.io/istio/install/kubernetes/helm/istio

Error: Could not get apiVersions from Kubernetes: unable to retrieve the complete list of server APIs: certmanager.k8s.io/v1alpha1: the server could not find the requested resource
```

**Comment by rlenglet** [comment]
@costinm @linsun it seems like this PR requires your approval?

**Pull request merged**